sycl/plugins/opencl/pi_opencl.cpp

```diff
@@ -71,19 +73,92 @@ CONSTFIX char clGetDeviceFunctionPointerName[] =
+    cl_program program, cl_uint spec_id, size_t spec_size,
+    const void *spec_value);
+
+struct ExtFuncsPerContextT;
+
+namespace detail {
+template <const char *FuncName, typename FuncT>
+std::pair<FuncT &, bool &> get(ExtFuncsPerContextT &);
+} // namespace detail
+
+struct ExtFuncsPerContextT {
+#define _EXT_FUNCTION_INTEL(t_pfx)                                             \
+  t_pfx##INTEL_fn t_pfx##Func = nullptr;                                       \
+  bool t_pfx##Initialized = false;
+
+#define _EXT_FUNCTION(t_pfx)                                                   \
+  t_pfx##_fn t_pfx##Func = nullptr;                                            \
+  bool t_pfx##Initialized = false;
+
+#include "ext_functions.inc"
+
+#undef _EXT_FUNCTION
+#undef _EXT_FUNCTION_INTEL
+
+  std::mutex Mtx;
+
+  template <const char *FuncName, typename FuncT>
+  std::pair<FuncT &, bool &> get() {
+    return detail::get<FuncName, FuncT>(*this);
+  }
+};
+
+namespace detail {
+
+#define _EXT_FUNCTION_COMMON(t_pfx, t_pfx_suff)                                \
+  template <>                                                                  \
+  std::pair<t_pfx_suff##_fn &, bool &> get<t_pfx##Name, t_pfx_suff##_fn>(      \
+      ExtFuncsPerContextT & Funcs) {                                           \
+    using FPtrT = t_pfx_suff##_fn;                                             \
+    std::pair<FPtrT &, bool &> Ret{Funcs.t_pfx##Func,                          \
+                                   Funcs.t_pfx##Initialized};                  \
+    return Ret;                                                                \
+  }
+#define _EXT_FUNCTION_INTEL(t_pfx) _EXT_FUNCTION_COMMON(t_pfx, t_pfx##INTEL)
+#define _EXT_FUNCTION(t_pfx) _EXT_FUNCTION_COMMON(t_pfx, t_pfx)
+
+#include "ext_functions.inc"
+
+#undef _EXT_FUNCTION
+#undef _EXT_FUNCTION_INTEL
+#undef _EXT_FUNCTION_COMMON
+} // namespace detail
+
+struct ExtFuncsCachesT {
+  std::map<pi_context, ExtFuncsPerContextT> Caches;
+  std::mutex Mtx;
+};
+
+ExtFuncsCachesT *ExtFuncsCaches = nullptr;
+
 // USM helper function to get an extension function pointer
 template <const char *FuncName, typename T>
 static pi_result getExtFuncFromContext(pi_context context, T *fptr) {
   // TODO
   // Potentially redo caching as PI interface changes.
-  thread_local static std::map<pi_context, T> FuncPtrs;
+  ExtFuncsPerContextT *PerContext = nullptr;
+  {
+    std::lock_guard<std::mutex> Lock{ExtFuncsCaches->Mtx};
+
+    PerContext = &ExtFuncsCaches->Caches[context];
+  }
+
+  std::lock_guard<std::mutex> Lock{PerContext->Mtx};
+  std::pair<T &, bool &> FuncInitialized = PerContext->get<FuncName, T>();
 
   // if cached, return cached FuncPtr
-  if (auto F = FuncPtrs[context]) {
+  if (FuncInitialized.second) {
     // if cached that extension is not available return nullptr and
     // PI_INVALID_VALUE
-    *fptr = F;
-    return F ? PI_SUCCESS : PI_INVALID_VALUE;
+    *fptr = FuncInitialized.first;
```

Comment:
Not sure we get a lot from having lazy initialization for these function pointers. Suggest moving all the initialization to `piContextCreate`. This should simplify the logic.

Reply:
Per what was discussed with @steffenlarsen , the main drawback of constant (as opposed to lazy) init in `piContextCreate` is that use-cases that create lots of contexts without using extensions will be impacted on performance. I reckon it's only synthetic cases, but still...

Reply:
I'm OK with leaving as is. But still think that if an application creates lots of context this initialization will be that last thing we should worry about. Also this would make "common" case faster and code more simpler because less mutexes and checks should be needed.